### ISSUE TITLE
[Question][Enhancement] Has anyone followed Dominic's attempt on the BBC?

### ISSUE THREAD

**Issue by Fabrizio-Caruso** [cc65/cc65#812]
Has anyone followed Dominic's attempt on the BBC?
https://github.com/dominicbeesley/cc65

How hard would it be to merge it? I see 691 commits.
It refers to branch 2.13. 

@polluks  Is this the BBC attempt you were refering in a different issue?
I ask because I have found other CC65 forks with some BBC support.
This one is semi-recent.

**Comment by polluks** [comment]
both :wink:

**Comment by Fabrizio-Caruso** [comment]
@polluks could you please elaborate more on this?
Do you think that a diff of branch 2.13 vs dominic's code is not applicable to the current master branch in a painless way?

**Comment by polluks** [comment]
@Fabrizio-Caruso It's on my TODO list. I was thinking about the exe format, if we should support these inf files, see http://archive.retro-kit.co.uk/bbc.nvg.org/std-format.php3.html.

**Comment by groessler** [comment]
I'm using some "Unix magic" to create the .inf file in my BBC assembler test program.
[bbc-test.zip](https://github.com/cc65/cc65/files/2787641/bbc-test.zip)

**Comment by Fabrizio-Caruso** [comment]
@polluks I am sending a letter to Santa to ask for some minimal BBC support. I would be more than happy even if it only produced a usable binary with no library support.

**Comment by groessler** [comment]
See my post from Jan-23. This is a "usable binary with no library support". An assembler program.
What exactly do you want? C should work, but no library support...

**Comment by Fabrizio-Caruso** [comment]
@groessler something like Supervision without joystick support which has no real library support (except for the joystick) but it comes with a linker's cfg file, a crt file, the binary produced is directly usable in an emulator and some macros definitions with some useful I/O addresses are available.
I will try your "unix magic" and see if I can build a hello world example.

**Comment by Fabrizio-Caruso** [comment]
@groessler googling bbcim has not been enough for me. I could not find a working link to the tool. :-( Do you have a working version of bbcim?

**Comment by polluks** [comment]
Well, I made a quick port for my system http://de.aminet.net/aminet/misc/emu/bbcim.readme

**Comment by groessler** [comment]
@polluks, where did you find the source code to bbcim? I didn't find the source code like Fabrizio, but I have a compiled version in my network.

**Comment by polluks** [comment]
@groessler here http://wouter.bbcmicro.net/bbc/bestanden/bbcim-1.2.tar.gz

**Comment by groessler** [comment]
Thanks!

**Comment by Fabrizio-Caruso** [comment]
@groessler your Makefile uses the option `-t bbc` which produces a missing `bbc.lib` error.
Should I use instead the `none` target and try to replicate bbc cfg memory layout?

**Comment by Fabrizio-Caruso** [comment]
@groessler How can I use your example with .c files instead of .s files? If I just use .c as you do with .s I get no .o file (just a .lst file)

**Comment by Fabrizio-Caruso** [comment]
@groessler Could you please provide an example with more than one object file and possible some non-target-specific math operations?
I am getting lots of errors. I understand that "-t bbc" cannot be used even for plain C because math libs are not found. I should probably use the none lib with bbc config. How can I do this cleanly?

I am currently doing the wrong thing by linking with:
```
	TINY_OBJ ?= character.o game_text.o level.o main.o move_player.o strategy.o
	$(CC65_PATH)/ld65$(EXE) -t bbc \
	--obj-path $(SOURCE_PATH) \
	-m $(BUILD_PATH)/bbc_no_graphics.map \
	-o $(BUILD_PATH)/bbc_no_graphics.bin \
	$(TINY_OBJ)
```
which produces:
```
ld65: Warning: /home/fabri/CC65_FIX/cc65/cfg/bbc.cfg(33): Segment 'STARTUP' does not exist
Unresolved external '_BOMB_IMAGE' referenced in:
  chase/level.s(204)
  chase/level.s(205)
Unresolved external '_CLEAR_SCREEN' referenced in:
  chase/main.s(136)
  chase/main.s(138)
  chase/main.s(140)
  chase/main.s(326)
  chase/main.s(84)
  chase/game_text.s(253)
Unresolved external '_DEAD_GHOST_IMAGE' referenced in:
...
Unresolved external 'tosultax' referenced in:
  chase/strategy.s(41)
  chase/character.s(454)
Unresolved external 'tosumulax' referenced in:
  chase/level.s(86)
  chase/level.s(187)
ld65: Error: 44 unresolved external(s) found - cannot create output file
make: *** [Makefile:1072: bbc_no_graphics] Error 1
```

**Comment by oliverschmidt** [comment]
@Fabrizio-Caruso: Do you still have an open question here?

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt, the answer is no longer open. I hope someone implements some minimal lib support one day. I will be more than willing to help in testing it.

**Comment by oliverschmidt** [comment]
Thanks for the clarification.

**Comment by Fabrizio-Caruso** [comment]
@polluks Any hope to see progress on this?
Are you still planning to implement some support for the bbc target? Have you made any progress or attempt?
As usual I am willing to help but my Assembly skills and understanding of how CC65 works are limited.
I would help with the tests and even lib code if I am able to.

The BBC Micro (and the related computers such as the Acorn Electron and the BBC Master) are the only well-known 6502-based target that is still missing in CC65.

We now have an alternative source of information to implement something here as VBCC does support the BBC (I have tried it successfully). I

**Comment by mrdudz** [comment]
Time to learn how to port the library to a new target :)

Step 1 is done already - there is a linker script that produces working asm binaries

now Step 2 is making a crt0 for the BBC.

For this i would do:
- pick a crt0 from another target (the none target perhaps, or if there is another that is "close" to the BBC use that perhaps)
- make this crt0 compile and link with an empty C program with the BBC target
- make a very simple C program that allows you to see if it works, like something that blinks the screen in a loop, or increments some character in screen memory

Once you have this, things should start working, and you can test if variables and the stack works correctly. And when you tweaked this, you "only" need to implement the target specific library.

**Comment by polluks** [comment]
@Fabrizio-Caruso Indeed, I have a real Elk and also some projects in mind.

**Comment by Fabrizio-Caruso** [comment]
@polluks LoL! An Elk? Meinst Du das Tier?? You mean the animal? Anyway, if you (and anyone else interested) would help with the crt and Assembly, I am willing to help on the BCC target. I could give it a try myself but I will be insulted for any single stupid Assembly line I write.
@mrdudz By the way... CC65 users' and devs' projects that use CC65 could be interesting for the others.
Would it make sense to present them in the Discussions section?

**Comment by polluks** [comment]
@Fabrizio-Caruso Elk and Beeb are the machine's nick name :-)
I took a closer look at Dominic's target: Nice features (runtime as language ROM, DFS support) but not very cc65-like (directory structure, etc.)

1. Top-down strategy: Make the whole target work with V2.19
2. Bottom-up strategy: Add each single feature to V2.19 (maybe a working conio is sufficient)

I prefer the latter cleaner strategy.

**Comment by mrdudz** [comment]
> By the way... CC65 users' and devs' projects that use CC65 could be interesting for the others.
Would it make sense to present them in the Discussions section?

Discussions are not suitable for collecting any sort of non volatile information imho - but we DO have such a list in the wiki already:

**Comment by Fabrizio-Caruso** [comment]
@mrdudz I will apply for having some one or two of my projects in the list: my own Cross-Lib and maybe a multiplexed sprite library for CC65 I developed with others.

**Comment by Fabrizio-Caruso** [comment]
@polluks I agree. Dominic's code (and maybe the code for the bbc target in vbcc might help, though). What is the most minimal, yet useful, support we can have to get it started? Having something like Supervision would be great (crt, cfg, joystick, maybe some macros and structs to access the hardware). Even less would be fine: just a crt. Once it is easy to use CC65 to produce a trivial example that display sa dot or a character, people (for example myself) will want to experiment with it.
There is already a cfg for the linker, which should probably already produce a binary that bbcim can convert into a .img file.

**Comment by Fabrizio-Caruso** [comment]
@mrdudz should we have a discussion for these feature request?

**Comment by mrdudz** [comment]
I have no idea (i just added some tags)

If you think he has interesting things in that fork - feel free to make a PR :)

**Comment by Fabrizio-Caruso** [comment]
@mrdudz That fork is too old and I am not able to bring it up to date and do a PR. That fork could be use for inspiration but it would require asm skills I don't have.
I was thinking of making a discussion for any work on the BCC target (which currently only exists in cc65 as a mere config file). The BBC is maybe only remaining target among the well-known computers.
